src/framework/components/esmscript/component.js

```diff
@@ -0,0 +1,446 @@
+    }
+
+    get isActive() {
+        return this.enabled && this.entity.enabled;
+    }
+
+    flushActiveModules() {
+
+        // ensure app-entity refs are up-to-date
+        this.appEntity.entity = this.entity;
+        this.appEntity.system = this.system;
+
+        for (const module of this.awaitingToBeEnabledModules) {
+            if (!this.isActive) break;
+            if (this.uninitializedModules.has(module)) continue;
+            this.awaitingToBeEnabledModules.delete(module);
+            this.enabledModules.add(module);
+            if (classHasMethod(module.constructor, 'update')) this.modulesWithUpdate.add(module);
+            if (classHasMethod(module.constructor, 'postUpdate')) this.modulesWithPostUpdate.add(module);
+            module.active?.(this.appEntity);
+        }
+    }
+
+    flushInactiveModules() {
+        for (const module of this.modules) {
+            if (!this.isActive) break;
+            if (this.enabledModules.has(module)) continue;
+            if (this.uninitializedModules.has(module)) continue;
+            module.inactive?.();
+        }
+    }
+
+    flushUninitializedModules() {
+        for (const module of this.uninitializedModules) {
+            if (!this.isActive) break;
+            if (!this.isModuleEnabled(module) && !this.awaitingToBeEnabledModules.has(module)) continue;
+            module.initialize();
+            this.uninitializedModules.delete(module);
+        }
+    }
+
+    onEnable() {
+        if (this.isActive) this.flushUninitializedModules();
+    }
+
+    _onInitialize() {
+        if (!this.initialized) this.flushUninitializedModules();
+        this.initialized = true;
+    }
+
+    _onPostInitialize() {
+        for (const module of this.modules) {
+            if (!this.isActive) break;
+            if (!this.enabledModules.has(module)) module.postInitialize?.();
+        }
+    }
+
+    _onUpdate(dt) {
+        for (const module of this.modulesWithUpdate) {
+            if (!this.isActive) break;
+            module.update(dt);
+        }
+    }
+
+    _onPostUpdate(dt) {
+        for (const module of this.modulesWithPostUpdate) {
+            if (!this.isActive) break;
+            module.postUpdate(dt);
+        }
+    }
+
+    /**
+     * Disables a module and prevents it receiving lifecycle events
+     * @param {ModuleInstance} module - The module to disable
+     */
+    disableModule(module) {
+
+        if (!this.modules.has(module)) {
+            Debug.error(`The ESM Script '${module?.constructor?.name}' has not been added to this component.`);
+            return;
+        }
+
+        this.enabledModules.delete(module);
+        this.awaitingToBeEnabledModules.delete(module);
+        this.modulesWithUpdate.delete(module);
+        this.modulesWithPostUpdate.delete(module);
+    }
+
+    /**
+     * Enables a module, allowing it to receive lifecycle events
+     * @param {ModuleInstance} module - The module to enable
+     * @internal
+     */
```

Comment:
In all jsdocs, we add full stop `.` at the end of the description.
And also an empty line after the description, before added @ tags. I don't remember the actual reason here, but I think the jsdocs might not get generated correctly .. try it.

Reply:
done